src/mixins/java/org/spongepowered/common/mixin/core/network/ConnectionMixin.java

```diff
@@ -148,7 +155,33 @@ public abstract class ConnectionMixin extends SimpleChannelInboundHandler<Packet
     @Inject(method = "lambda$doSendPacket$8", at = @At(value = "INVOKE", target = "Lio/netty/util/concurrent/Future;isSuccess()Z"))
     public void impl$onPacketSent(final PacketSendListener $$0x, final Future $$1x, final CallbackInfo ci) {
         if ($$0x instanceof final PacketSender.SpongePacketSendListener spongeListener) {
-            spongeListener.accept($$1x);
+            spongeListener.accept($$1x.cause());
+        }
+    }
+
+    @Inject(method = "send(Lnet/minecraft/network/protocol/Packet;Lnet/minecraft/network/PacketSendListener;)V", at = @At(value = "HEAD"), cancellable = true)
+    private void impl$onSend(final Packet<?> $$0, final @Nullable PacketSendListener $$1, final CallbackInfo ci) {
+        if (this.disconnectionHandled) {
+            if ($$1 instanceof final PacketSender.SpongePacketSendListener spongeListener) {
+                spongeListener.accept(new IOException("Connection has been closed."));
+            }
+
+            ci.cancel();
```

Comment:
Cancelling packet send when disconnected sounds logic to me but vanilla doesn't do that. I don't know what the implications are.

Reply:
Vanilla doesn't really have any concerns but mods might but that seems highly unlikely to me. The connection is effectively closed at this point and we are just piling up the packets in the `#queue` field which doesn't get touched after disconnection. So this is pretty much just saving up some allocations.

Reply:
Alright, that should be fine then.